### ISSUE TITLE
Playwright: Investigate migration path.

### ISSUE THREAD

**Issue by worldomonation** [Automattic/wp-calypso#50693]
Topic came up in biweekly team meeting for .com quality squad.

Due to the sheer amount of work required to shore up flakey e2e tests based around Selenium, consider spending time investigating whether it is possible for existing infrastructure to support both WebdriverJS and Playwright tests in parallel.

**Comment by worldomonation** [comment]
Couple of things:

1. getting tests to run under Playwright requires some scaffolding and stubbing, but it is a relatively straightforward task. This is illustrated by my trial project at #47874. The amount of scaffolding is not insignificant, but because Playwright is much more efficient and reliable at waiting for selectors and clicking them, the amount of code decreases compared to Selenium.
2. the more significant challenge is to how to keep existing Selenium test and infrastructure around, while developing the Playwright tests and migrating them over. There are couple of ways I can think of off top my head:

a) have a parallel file that is migrated to Playwright. Playwright tests will call these files instead of the original Selenium based files. For example, during the migration phase there may be two `LoginFlow` files in the repo, where one is the existing Selenium based file and the other being the Playwright based file. 

This approach has advantage in that there will be minimal disruption to existing Selenium tests and the new Playwright infrastructure will run quietly in the background.

The challenge with this approach is that temporarily, we have to manage two sets of files, and once the migration is complete all references to the then-deprecated Selenium based files need to be removed from the repository, a tedious task. 

b) support both codepaths in same file. Code for both frameworks will live in the same file, just distinguished via a switch statement or a keyword that defaults to 'Selenium'.

This approach means there is less of a clean-up and shuffling of files around, as all code is contained in the same file as existing infrastructure. it also results in less files being created and deleted, and will keep revision history in GitHub as well.

The disadvantage of this method is that files will churn a lot, especially if other developers work on the file at the same time. It will also lead to inevitable complexity and mess as various if/else statements are added in a bunch of places. This approach will also result in existing file structure being kept, even if it may not make sense for Playwright-based paradigm.

----

In the end, I will have to try out some proof of concept of the 2b) as I already have experience implementing 2a) as part of the trial.

**Comment by worldomonation** [comment]
Some more days spent at the helm of this issue, the status report is as follows (as of March 5):

- efforts were made to have the `playwright` demo work on CircleCI. However, this was unsuccessful due to issues with CircleCI and `xvfb`; specifically the error is that `xvfb cannot open display`. This is a common issue for those running a flavor of Linux, but not a common error on CircleCI. After a day of trying to make it work, I have decided to concentrate my efforts on getting things to work with TeamCity instead.
- attempts to run demo script on TeamCity was hampered by several factors, including the non-persistence of the location where `playwright` downloaded binaries on the disk. This meant that between the environment setup step and the actual test execution step, these binaries were removed from the system resulting in `playwright` not being able to locate them when the demo script was called. Generally speaking, my lack of experience with TeamCity is hampering progress, but I am learning the system as I go.

As it currently stands, these are the action items, in no particular order:

1. get the `playwright` browser binaries to install into a directory that persists between steps.
2. request administrator privilege for the Calypso TeamCity project in order to set up an isolated test project.
3. set up a docker image that has `playwright` pre-installed (if necessary or beneficial)
4. develop a test script that calls `playwright` and run on TeamCity.
5. devise a plan for migrating test scripts to `playwright`, including where the test will live, how it will be run, whether `playwright` tests will have its own run instead, test runner of choice, etc.

**Comment by worldomonation** [comment]
One item done: getting `playwright` to install and run in the TeamCity environment.

Made a request on Systems P2 for administrator access for TeamCity as well as a new project.

**Comment by worldomonation** [comment]
#50774 has been merged into trunk.

The PR above installs Playwright as one of the dependencies in `test/e2e` and lays the groundwork for actually migrating over a demonstration script.

**Comment by worldomonation** [comment]
Additional branches are awaiting PR to be submitted to reviewers, and also to rebase on top of #51149.

**Comment by worldomonation** [comment]
Done for now.
